### ISSUE TITLE
DiceMetric refactoring

### ISSUE THREAD

**Issue by myron** [Project-MONAI/MONAI#5378]
Dice metric is probably the most popular accuracy metric used for medical images.  

Our current implementation is lacking features, slow and bug-prone since it's very convoluted coding. 
https://github.com/Project-MONAI/MONAI/blob/dev/monai/metrics/meandice.py#L22

Issues

1) Long subclass chain: DiceMetric  class -> CumulativeIterationMetric class -> Cumulative class + IterationMetric class -> Metric class. For something simple to compute such as Dice, it takes time understand it or debug 

2) The implementation is not intuitive, when calling class instance
For example
```
metric = DiceMetric(reduction = 'mean')
dice = metric(x,y)
```
Usually one would expect a mean dice value, instead we return an un-reduced  result. Futhermore we add intermediate result to "buffers", which is hidden from the the user and non-obvious
to get my mean dice, I'll need to do these steps
```
metric = DiceMetric(reduction = 'mean')
metric(x,y)
dice = metric.aggregate()
```
but even worse, if within my training loop I call this metric several times, I need to reset the buffers manually:
```
metric = DiceMetric(reduction = 'mean')
for i in range(100):
  metric(x[i],y[i])
  dice = metric.aggregate()
  metric.reset()
```
If we don't reset , then the buffer will keep accumulating data, and the aggregate result will be an overall aggregate (not from the current images).   But then we have a second issue, if we .reset() then we loose the ability to find  the total mean dice at the end of the loop.  

----
3) The hidden buffers.  with each call to DiceMetric() instance, we keep adding intermediate results (BatchxChannels) to a buffer (usually on GPU) which growth with each call.  
And we  return the cloned buffers 

https://github.com/Project-MONAI/MONAI/blob/c27ab360212ad122f5ca41fc69b8c2f47676b809/monai/metrics/metric.py#L269

So we're growing gpu memory unnecessarily, and cloning at each aggregate call. 

And in DDP (on multi gpu) it's even slower
https://github.com/Project-MONAI/MONAI/blob/c27ab360212ad122f5ca41fc69b8c2f47676b809/monai/metrics/metric.py#L253

we gather all buffers from all gpus,  this requires Allocating new gpu memory (the same size as buffers) to synchronize.
The syncronization itself is using our own function evenly_divisible_all_gather, which adds further overhead. 

the size of gpu beffers per gpu is multiplicative of the number of iterations / images in the dataset. So for larger dataset this is not a negligible gpu memory overhead.

4) We don't really need accumulating buffers (only averages),  and since we accumulate buffers,  in multi-gpu with non-evenly divisible by n_gpus data length, we can end up with buffers of different length between the gpus, and then we need to worry how to reduce it properly without the dead-lock. The evenly_divisible_all_gather might be solving it with an overhead. But all this effort is unnecessary.

5) Missing features. For the companion loss function  DiceLoss(softmax=True, to_onehot_y=True) we can specify extract options (softmax, to_onehot_y)   so that we can call it on the output of the network. But DiceMetric, we don't have it, and a user needs to do all the pre-processing. I think we should be consistent add these options (it's the easiest to solve)

Overall, I think we should simplify and refactor DiceMetric calculation, it should a simple class that compute mean dice (or provided reduction).  Currently the internal logic is so complicated that it's very difficult to debug or even understand.  The practical use is also inconvenient (requires a lot of code lines) and non-intuitive. 

I recommend simplify it and even better separate it from all super classes. Perhaps we can simply return the result, and multi-gpu syncronization can be in a separate class (if needed). Furthermore we don't need these gpu buffers and can simply track the intermediate average.   Alternatively we can keep this as is, and add an additional DiceMetricSimple() class for simple and straightforward dice calculation.

**Comment by myron** [comment]
Here is draft PR with 2 variants: further modify the existing DiceMetric() class (it becomes even more convoluted) or create a  new simple DiceMetricSimple() class https://github.com/Project-MONAI/MONAI/pull/5379

**Comment by myron** [comment]
Here is another PR that adds a general purpose utility to calculate average accumulated values, and can be used  (for this issue too) to disentangle metric computation from the multi-gpu reduction handling. https://github.com/Project-MONAI/MONAI/pull/5380
